python/python/lance/arrow.py

```diff
@@ -11,14 +11,347 @@
 #  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
-from typing import Optional, Union
+from pathlib import Path
+from typing import Iterable, Optional, Union
 
 import pyarrow as pa
 
 from .lance import BFloat16
 from .lance import bfloat16_array as bfloat16_array
 
 
+# TODO: add VariableShapeImageTensorType once pa.VariableShapeTensorArray is available
+class ImageURIType(pa.ExtensionType):
+    def __init__(self):
+        pa.ExtensionType.__init__(self, pa.string(), "lance.arrow.image_uri")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return ImageURIType()
+
+    def __arrow_ext_class__(self):
+        return ImageURIArray
+
+    def __arrow_ext_scalar_class__(self):
+        return ImageURIScalar
+
+
+class EncodedImageType(pa.ExtensionType):
+    def __init__(self):
+        # TODO: use pa.BinaryView once available?
+        pa.ExtensionType.__init__(self, pa.binary(), "lance.arrow.encoded_image")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return EncodedImageType()
+
+    def __arrow_ext_class__(self):
+        return EncodedImageArray
+
+    def __arrow_ext_scalar_class__(self):
+        return EncodedImageScalar
+
+
+class FixedShapeImageTensorType(pa.ExtensionType):
+    def __init__(self, arrow_type, shape):
+        pa.ExtensionType.__init__(
+            self,
+            pa.fixed_shape_tensor(arrow_type, shape),
```

Comment:
Is it legal for the storage type to be an extension type? Wondering if this could cause issues down the road.

Reply:
There is currently an issue serializing `FixedShapeImageTensorArray` and I'm not sure where it's caused, but I'll have to sort it out tomorrow.

Reply:
I suppose non canonical types won't work with arrow-rs so calling [ArrowArrayStreamReader::from_pyarrow](https://github.com/apache/arrow-rs/blob/master/arrow/src/pyarrow.rs#L283-L302).

Reply:
SSwitched to subclassing `pa.FixedSizeList` so this is no longer relavant.